Source/OrchBase/Placement/CostCache.cpp

```diff
@@ -0,0 +1,611 @@
+    {
+        /* For each board edge (in each direction)... */
+        P_board* fromBoard;
+        P_board* toBoard;
+
+        WALKPDIGRAPHARCS(AddressComponent, P_board*,
+                         unsigned, P_link*,
+                         unsigned, P_port*, engine->G, boardEdge)
+        {
+            /* Get the boards. */
+            fromBoard = boardEdge->second.fr_n->second.data;
+            toBoard = boardEdge->second.to_n->second.data;
+
+
+            /* Grab each mailbox pairing */
+            P_mailbox* fromMailbox;
+            P_mailbox* toMailbox;
+            bool join;
+
+            WALKPDIGRAPHNODES(AddressComponent, P_mailbox*,
+                              unsigned, P_link*,
+                              unsigned, P_port*,
+                              fromBoard->G, mailboxFromNode)
+            WALKPDIGRAPHNODES(AddressComponent, P_mailbox*,
+                              unsigned, P_link*,
+                              unsigned, P_port*,
+                              toBoard->G, mailboxToNode)
+            {
+                fromMailbox = mailboxFromNode->second.data;
+                toMailbox = mailboxToNode->second.data;
+                join = false;
+
+                /* N->S */
+                if (fromBoard->Name()[2] == '1' and toBoard->Name()[2] == '0')
+                {
+                    if (fromMailbox->Name()[2] == '0' and
+                        toMailbox->Name()[2] == '3' and
+                        fromMailbox->Name()[1] == toMailbox->Name()[1]) join = true;
+                }
+
+                /* S->N */
+                else if (fromBoard->Name()[2] == '0' and toBoard->Name()[2] == '1')
+                {
+                    if (fromMailbox->Name()[2] == '3' and
+                        toMailbox->Name()[2] == '0' and
+                        fromMailbox->Name()[1] == toMailbox->Name()[1]) join = true;
+                }
+
+                /* E->W */
+                else if ((fromBoard->Name()[1] == '0' and toBoard->Name()[1] == '1') or
+                         (fromBoard->Name()[1] == '1' and toBoard->Name()[1] == '2'))
+                {
+                    if (fromMailbox->Name()[1] == '3' and
+                        toMailbox->Name()[1] == '0' and
+                        fromMailbox->Name()[2] == toMailbox->Name()[2]) join = true;
+                }
+
+                /* W->E */
+                else if ((fromBoard->Name()[1] == '1' and toBoard->Name()[1] == '0') or
+                         (fromBoard->Name()[1] == '2' and toBoard->Name()[1] == '1'))
+                {
+                    if (fromMailbox->Name()[1] == '0' and
+                        toMailbox->Name()[1] == '3' and
+                        fromMailbox->Name()[2] == toMailbox->Name()[2]) join = true;
+                }
+
+                else
+                {
+                    printf("Couldn't determine relationship between boards "
```

Comment:
logserver?

Reply:
I've addressed this - this bit of control flow now throws, and is handled by the caller. Two new logserver messages and a new test case for the reader.